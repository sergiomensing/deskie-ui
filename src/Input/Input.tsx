import { ChangeEvent, forwardRef, InputHTMLAttributes, useImperativeHandle, useRef } from 'react';
import styled, { css } from 'styled-components';

export type InputStatus = 'default' | 'success' | 'warning' | 'error';

type BaseInputProps = {
  icon?: React.ReactNode;
  iconRight?: React.ReactNode;
  status?: InputStatus;
  onChange?: (ev: ChangeEvent<HTMLInputElement>) => void;
};

export type InputProps = BaseInputProps & Omit<InputHTMLAttributes<unknown>, keyof BaseInputProps>;

const StyledInput = styled.div<{ icon: boolean; iconRight: boolean; disabled: boolean; status: InputStatus }>`
  display: flex;
  align-items: center;
  position: relative;

  input {
    width: 100%;
    display: flex;
    height: 3.5rem;
    padding: 0 0.75rem;
    font-weight: 400;
    font: inherit;
    border: 1px solid
      ${(props) => (props.status === 'default' ? props.theme.colors.accent8 : props.theme.colors[props.status])};
    border-radius: 0.5rem;
    box-shadow: none;
    background: ${(props) => props.theme.colors.background};
    -webkit-appearance: none;
    color: inherit;
    cursor: pointer;
    position: relative;
    z-index: 1;

    ${(props) => props.icon && `padding-left: 2.25rem;`};
    ${(props) => props.iconRight && `padding-right: 2.25rem;`};

    ${(props) =>
      props.disabled &&
      css`
        cursor: not-allowed;
        border: 1px solid ${props.theme.colors.accent6};
        color: ${props.theme.colors.accent7};
      `}

    &:focus {
      border-color: ${(props) => props.theme.colors.primary};
      box-shadow: 0px 0px 0px 1px ${(props) => props.theme.colors.primary};
      outline: 0;
    }
  }

  .input-icon {
    position: absolute;
    z-index: 2;
    width: 1.25rem;
    left: 0.75rem;

    svg {
      width: 100%;
      height: auto;
    }
  }

  .input-icon-right {
    left: initial;
    right: 0.75rem;
  }
`;

const Input = forwardRef<HTMLInputElement, InputProps>(
  (
    { disabled = false, icon, iconRight, status = 'default', className = '', ...rest },
    ref: React.Ref<HTMLInputElement | null>,
  ) => {
    const inputRef = useRef<HTMLInputElement>(null);
    useImperativeHandle(ref, () => inputRef.current);

    return (
      <StyledInput
        className={`input-container ${className}`}
        disabled={disabled}
        icon={Boolean(icon)}
        status={status}
        iconRight={Boolean(iconRight)}
      >
        {icon && <span className="input-icon">{icon}</span>}
        <input ref={inputRef} disabled={disabled} {...rest} />
        {iconRight && <span className="input-icon input-icon-right">{iconRight}</span>}
      </StyledInput>
    );
  },
);

export default Input;
