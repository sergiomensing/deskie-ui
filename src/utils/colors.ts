export const colorShade = (col: string, amt: number) => {
  col = col.replace(/^#/, '');
  if (col.length === 3) col = col[0] + col[0] + col[1] + col[1] + col[2] + col[2];

  const [sr, sg, sb] = col.match(/.{2}/g) || ['ff', 'ff', 'ff'];
  const [ar, ag, ab] = [parseInt(sr, 16) + amt, parseInt(sg, 16) + amt, parseInt(sb, 16) + amt];

  const r = Math.max(Math.min(255, ar), 0).toString(16);
  const g = Math.max(Math.min(255, ag), 0).toString(16);
  const b = Math.max(Math.min(255, ab), 0).toString(16);

  const rr = (r.length < 2 ? '0' : '') + r;
  const gg = (g.length < 2 ? '0' : '') + g;
  const bb = (b.length < 2 ? '0' : '') + b;

  return `#${rr}${gg}${bb}`;
};

const defaultColors = [
  '#E64980',
  '#4C6EF5',
  '#228BE6',
  '#15AABF',
  '#12B886',
  '#40C057',
  '#82C91E',
  '#FAB005',
  '#FD7E14',
  '#FA5252',
];

export function getColorFromString(value: string, colors: undefined | string[] = defaultColors) {
  let hash = 0;
  if (value.length === 0) return colors[hash];
  for (let i = 0; i < value.length; i++) {
    hash = value.charCodeAt(i) + ((hash << 5) - hash);
    hash = hash & hash;
  }
  hash = ((hash % colors.length) + colors.length) % colors.length;
  return colors[hash];
}
