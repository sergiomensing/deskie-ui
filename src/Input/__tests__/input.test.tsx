import React from 'react';
import Input from '../Input';
import { fireEvent, render } from '../../../tests/utils';

describe('Input', () => {
  it('should support different statuses', () => {
    const { container } = render(
      <>
        <Input status="default" />
        <Input status="error" />
        <Input status="success" />
        <Input status="warning" />
      </>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render icons', () => {
    const { getByTestId, container } = render(
      <>
        <Input icon={<span data-testid="test-icon"></span>} />
        <Input iconRight={<span data-testid="test-icon-right"></span>} />
      </>,
    );

    expect(getByTestId('test-icon')).toBeInTheDocument();
    expect(getByTestId('test-icon-right')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should trigger event when input changed', () => {
    const Wrapper = () => {
      const [state, setState] = React.useState<string>('test-value1');

      return <Input data-testid="test-input" value={state} onChange={(e) => setState(e.target.value)} />;
    };

    const { getByTestId } = render(<Wrapper />);

    const input = getByTestId('test-input');

    expect(input).toHaveValue('test-value1');

    fireEvent.change(input, { target: { value: 'test-value2' } });

    expect(input).toHaveValue('test-value2');
  });

  it('should forward ref', () => {
    const ref = React.createRef<HTMLInputElement>();
    const { container } = render(<Input readOnly value="test-value" ref={ref} />);
    expect(container.querySelector('input')).toEqual(ref.current);
  });
});
