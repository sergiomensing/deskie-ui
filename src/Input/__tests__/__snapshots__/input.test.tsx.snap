// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render icons 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
}

.c0 input {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 3.5rem;
  padding: 0 0.75rem;
  font-weight: 400;
  font: inherit;
  border: 1px solid #9395A1;
  border-radius: 0.5rem;
  box-shadow: none;
  background: #FFFFFF;
  -webkit-appearance: none;
  color: inherit;
  cursor: pointer;
  position: relative;
  z-index: 1;
  padding-left: 2.25rem;
}

.c0 input:focus {
  border-color: #6C5DD3;
  box-shadow: 0px 0px 0px 1px #6C5DD3;
  outline: 0;
}

.c0 .input-icon {
  position: absolute;
  z-index: 2;
  width: 1.25rem;
  left: 0.75rem;
}

.c0 .input-icon svg {
  width: 100%;
  height: auto;
}

.c0 .input-icon-right {
  left: initial;
  right: 0.75rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
}

.c1 input {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 3.5rem;
  padding: 0 0.75rem;
  font-weight: 400;
  font: inherit;
  border: 1px solid #9395A1;
  border-radius: 0.5rem;
  box-shadow: none;
  background: #FFFFFF;
  -webkit-appearance: none;
  color: inherit;
  cursor: pointer;
  position: relative;
  z-index: 1;
  padding-right: 2.25rem;
}

.c1 input:focus {
  border-color: #6C5DD3;
  box-shadow: 0px 0px 0px 1px #6C5DD3;
  outline: 0;
}

.c1 .input-icon {
  position: absolute;
  z-index: 2;
  width: 1.25rem;
  left: 0.75rem;
}

.c1 .input-icon svg {
  width: 100%;
  height: auto;
}

.c1 .input-icon-right {
  left: initial;
  right: 0.75rem;
}

<div>
  <div
    class="c0 input-container "
  >
    <span
      class="input-icon"
    >
      <span
        data-testid="test-icon"
      />
    </span>
    <input />
  </div>
  <div
    class="c1 input-container "
  >
    <input />
    <span
      class="input-icon input-icon-right"
    >
      <span
        data-testid="test-icon-right"
      />
    </span>
  </div>
</div>
`;

exports[`Input should support different statuses 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
}

.c0 input {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 3.5rem;
  padding: 0 0.75rem;
  font-weight: 400;
  font: inherit;
  border: 1px solid #9395A1;
  border-radius: 0.5rem;
  box-shadow: none;
  background: #FFFFFF;
  -webkit-appearance: none;
  color: inherit;
  cursor: pointer;
  position: relative;
  z-index: 1;
}

.c0 input:focus {
  border-color: #6C5DD3;
  box-shadow: 0px 0px 0px 1px #6C5DD3;
  outline: 0;
}

.c0 .input-icon {
  position: absolute;
  z-index: 2;
  width: 1.25rem;
  left: 0.75rem;
}

.c0 .input-icon svg {
  width: 100%;
  height: auto;
}

.c0 .input-icon-right {
  left: initial;
  right: 0.75rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
}

.c1 input {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 3.5rem;
  padding: 0 0.75rem;
  font-weight: 400;
  font: inherit;
  border: 1px solid #FF754C;
  border-radius: 0.5rem;
  box-shadow: none;
  background: #FFFFFF;
  -webkit-appearance: none;
  color: inherit;
  cursor: pointer;
  position: relative;
  z-index: 1;
}

.c1 input:focus {
  border-color: #6C5DD3;
  box-shadow: 0px 0px 0px 1px #6C5DD3;
  outline: 0;
}

.c1 .input-icon {
  position: absolute;
  z-index: 2;
  width: 1.25rem;
  left: 0.75rem;
}

.c1 .input-icon svg {
  width: 100%;
  height: auto;
}

.c1 .input-icon-right {
  left: initial;
  right: 0.75rem;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
}

.c2 input {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 3.5rem;
  padding: 0 0.75rem;
  font-weight: 400;
  font: inherit;
  border: 1px solid #7FBA7A;
  border-radius: 0.5rem;
  box-shadow: none;
  background: #FFFFFF;
  -webkit-appearance: none;
  color: inherit;
  cursor: pointer;
  position: relative;
  z-index: 1;
}

.c2 input:focus {
  border-color: #6C5DD3;
  box-shadow: 0px 0px 0px 1px #6C5DD3;
  outline: 0;
}

.c2 .input-icon {
  position: absolute;
  z-index: 2;
  width: 1.25rem;
  left: 0.75rem;
}

.c2 .input-icon svg {
  width: 100%;
  height: auto;
}

.c2 .input-icon-right {
  left: initial;
  right: 0.75rem;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
}

.c3 input {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 3.5rem;
  padding: 0 0.75rem;
  font-weight: 400;
  font: inherit;
  border: 1px solid #FFCE73;
  border-radius: 0.5rem;
  box-shadow: none;
  background: #FFFFFF;
  -webkit-appearance: none;
  color: inherit;
  cursor: pointer;
  position: relative;
  z-index: 1;
}

.c3 input:focus {
  border-color: #6C5DD3;
  box-shadow: 0px 0px 0px 1px #6C5DD3;
  outline: 0;
}

.c3 .input-icon {
  position: absolute;
  z-index: 2;
  width: 1.25rem;
  left: 0.75rem;
}

.c3 .input-icon svg {
  width: 100%;
  height: auto;
}

.c3 .input-icon-right {
  left: initial;
  right: 0.75rem;
}

<div>
  <div
    class="c0 input-container "
  >
    <input />
  </div>
  <div
    class="c1 input-container "
  >
    <input />
  </div>
  <div
    class="c2 input-container "
  >
    <input />
  </div>
  <div
    class="c3 input-container "
  >
    <input />
  </div>
</div>
`;
